File: src-tauri/src/events.rs
// SPDX-License-Identifier: MPL-2.0

//! Typed event definitions and emit helpers.
//!
//! All backend → frontend events are defined here. Command handlers and background
//! tasks use `emit_*` helpers rather than calling `app_handle.emit()` directly,
//! ensuring the event name and payload type are always in sync.
//!
//! Event names are defined by the `#[tauri_specta(event_name = "...")]` attribute
//! on each event struct. The helpers below use `tauri_specta::Event::emit()`.

pub mod types;

pub use types::*;

use tauri::AppHandle;
use tauri_specta::Event;

/// Emit a session topology change event.
pub fn emit_session_state_changed(app: &AppHandle, event: SessionStateChangedEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit session-state-changed: {e}");
    }
}

/// Emit an SSH lifecycle state change event.
pub fn emit_ssh_state_changed(app: &AppHandle, event: SshStateChangedEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit ssh-state-changed: {e}");
    }
}

/// Emit a screen update event (dirty cells from VT processing).
pub fn emit_screen_update(app: &AppHandle, event: ScreenUpdateEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit screen-update: {e}");
    }
}

/// Emit a terminal mode change event (DECCKM / DECKPAM).
pub fn emit_mode_state_changed(app: &AppHandle, event: ModeStateChangedEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit mode-state-changed: {e}");
    }
}

/// Emit a scrollback position change event.
pub fn emit_scroll_position_changed(app: &AppHandle, event: ScrollPositionChangedEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit scroll-position-changed: {e}");
    }
}

/// Emit a credential prompt event (SSH auth needs user input).
pub fn emit_credential_prompt(app: &AppHandle, event: CredentialPromptEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit credential-prompt: {e}");
    }
}

/// Emit a passphrase prompt event (SSH pubkey auth — encrypted private key, FS-SSH-019a).
pub fn emit_passphrase_prompt(app: &AppHandle, event: PassphrasePromptEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit passphrase-prompt: {e}");
    }
}

/// Emit a host key prompt event (TOFU verification).
pub fn emit_host_key_prompt(app: &AppHandle, event: HostKeyPromptEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit host-key-prompt: {e}");
    }
}

/// Emit a pane notification change event (bell, background output, process exit).
pub fn emit_notification_changed(app: &AppHandle, event: NotificationChangedEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit notification-changed: {e}");
    }
}

/// Emit a cursor style change event (DECSCUSR).
pub fn emit_cursor_style_changed(app: &AppHandle, event: CursorStyleChangedEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit cursor-style-changed: {e}");
    }
}

/// Emit a bell triggered event (rate-limited, at most 1 per 100 ms per pane).
pub fn emit_bell_triggered(app: &AppHandle, event: BellTriggeredEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit bell-triggered: {e}");
    }
}

/// Emit an OSC 52 clipboard write request event (FS-VT-075).
pub fn emit_osc52_write_requested(app: &AppHandle, event: Osc52WriteRequestedEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit osc52-write-requested: {e}");
    }
}

/// Emit a deprecated SSH algorithm warning event (FS-SSH-014).
pub fn emit_ssh_warning(app: &AppHandle, event: SshWarningEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit ssh-warning: {e}");
    }
}

/// Emit an SSH reconnected separator event (FS-SSH-042).
pub fn emit_ssh_reconnected(app: &AppHandle, event: SshReconnectedEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit ssh-reconnected: {e}");
    }
}

/// Emit a full-screen state change event (FS-FULL-009).
pub fn emit_fullscreen_state_changed(app: &AppHandle, event: FullscreenStateChangedEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit fullscreen-state-changed: {e}");
    }
}

/// Emit a preferences-changed event (cross-instance sync).
pub fn emit_preferences_changed(app: &AppHandle, event: PreferencesChangedEvent) {
    if let Err(e) = event.emit(app) {
        tracing::error!("Failed to emit preferences-changed: {e}");
    }
}

File: src/lib/ipc/events.ts
// SPDX-License-Identifier: MPL-2.0

/**
 * Typed wrappers for all IPC event subscriptions (listen() calls).
 *
 * Each function returns an unsubscribe callback (Promise<() => void>) that
 * must be called on component destroy to prevent memory leaks.
 *
 * Event names mirror the Rust backend emit calls in src-tauri/src/events/.
 */

import { listen } from '@tauri-apps/api/event';
import type {
  SessionStateChangedEvent,
  SshStateChangedEvent,
  SshWarningEvent,
  SshReconnectedEvent,
  HostKeyPromptEvent,
  CredentialPromptEvent,
  PassphrasePromptEvent,
  NotificationChangedEvent,
  ModeStateChangedEvent,
  ScreenUpdateEvent,
  ScrollPositionChangedEvent,
  CursorStyleChangedEvent,
  BellTriggeredEvent,
  FullscreenStateChangedEvent,
} from './types';

// ---------------------------------------------------------------------------
// Session events
// ---------------------------------------------------------------------------

export function onSessionStateChanged(
  handler: (event: SessionStateChangedEvent) => void,
): Promise<() => void> {
  return listen<SessionStateChangedEvent>('session-state-changed', (e) => handler(e.payload));
}

// ---------------------------------------------------------------------------
// SSH events
// ---------------------------------------------------------------------------

export function onSshStateChanged(
  handler: (event: SshStateChangedEvent) => void,
): Promise<() => void> {
  return listen<SshStateChangedEvent>('ssh-state-changed', (e) => handler(e.payload));
}

export function onSshWarning(handler: (event: SshWarningEvent) => void): Promise<() => void> {
  return listen<SshWarningEvent>('ssh-warning', (e) => handler(e.payload));
}

export function onSshReconnected(
  handler: (event: SshReconnectedEvent) => void,
): Promise<() => void> {
  return listen<SshReconnectedEvent>('ssh-reconnected', (e) => handler(e.payload));
}

export function onHostKeyPrompt(handler: (event: HostKeyPromptEvent) => void): Promise<() => void> {
  return listen<HostKeyPromptEvent>('host-key-prompt', (e) => handler(e.payload));
}

export function onCredentialPrompt(
  handler: (event: CredentialPromptEvent) => void,
): Promise<() => void> {
  return listen<CredentialPromptEvent>('credential-prompt', (e) => handler(e.payload));
}

export function onPassphrasePrompt(
  handler: (event: PassphrasePromptEvent) => void,
): Promise<() => void> {
  return listen<PassphrasePromptEvent>('passphrase-prompt', (e) => handler(e.payload));
}

// ---------------------------------------------------------------------------
// Notification events
// ---------------------------------------------------------------------------

export function onNotificationChanged(
  handler: (event: NotificationChangedEvent) => void,
): Promise<() => void> {
  return listen<NotificationChangedEvent>('notification-changed', (e) => handler(e.payload));
}

// ---------------------------------------------------------------------------
// Terminal mode events
// ---------------------------------------------------------------------------

export function onModeStateChanged(
  handler: (event: ModeStateChangedEvent) => void,
): Promise<() => void> {
  return listen<ModeStateChangedEvent>('mode-state-changed', (e) => handler(e.payload));
}

// ---------------------------------------------------------------------------
// Screen update events
// ---------------------------------------------------------------------------

export function onScreenUpdate(handler: (event: ScreenUpdateEvent) => void): Promise<() => void> {
  return listen<ScreenUpdateEvent>('screen-update', (e) => handler(e.payload));
}

export function onScrollPositionChanged(
  handler: (event: ScrollPositionChangedEvent) => void,
): Promise<() => void> {
  return listen<ScrollPositionChangedEvent>('scroll-position-changed', (e) => handler(e.payload));
}

export function onCursorStyleChanged(
  handler: (event: CursorStyleChangedEvent) => void,
): Promise<() => void> {
  return listen<CursorStyleChangedEvent>('cursor-style-changed', (e) => handler(e.payload));
}

export function onBellTriggered(handler: (event: BellTriggeredEvent) => void): Promise<() => void> {
  return listen<BellTriggeredEvent>('bell-triggered', (e) => handler(e.payload));
}

// ---------------------------------------------------------------------------
// Fullscreen events
// ---------------------------------------------------------------------------

export function onFullscreenStateChanged(
  handler: (event: FullscreenStateChangedEvent) => void,
): Promise<() => void> {
  return listen<FullscreenStateChangedEvent>('fullscreen-state-changed', (e) => handler(e.payload));
}
